src/helpers/fabric.rs

```diff
@@ -0,0 +1,56 @@
+use crate::helpers::error::Error;
+use crate::helpers::Identity;
+use crate::protocol::{RecordId, Step};
+use async_trait::async_trait;
+use futures::Stream;
+use std::fmt::{Debug, Formatter};
+
+/// Combination of helper identity and step that uniquely identifies a single channel of communication
+/// between two helpers.
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct ChannelId<S> {
+    pub identity: Identity,
+    pub step: S,
+}
+
+#[derive(Debug)]
+pub struct MessageEnvelope {
+    pub record_id: RecordId,
+    pub payload: Box<[u8]>,
+}
+
+pub type MessageChunks<S> = (ChannelId<S>, Vec<MessageEnvelope>);
+
+#[async_trait]
+pub trait Fabric<S: Step>: Sync {
+    /// Type of the channel that is used to send messages to other helpers
+    type Channel: CommunicationChannel;
+    type MessageStream: Stream<Item = MessageChunks<S>> + Send + Unpin + 'static;
+
+    /// Returns a new connection to be open. `channel_id` indicates the parameters of this
+    /// connection (destination peer and step). Once the connection is returned it is immediately
+    /// ready for sending messages.
+    async fn get_connection(&self, channel_id: ChannelId<S>) -> Self::Channel;
+
+    /// Returns a stream to receive messages that have arrived from other helpers. Note that
+    /// some implementations may panic if this method is called more than once.
+    fn message_stream(&self) -> Self::MessageStream;
```

Comment:
Would it be better to construct this as a pair of objects instead?  One that you can hand out for sending purposes and another for handling the messages from other peers?

(It's possible that what we really want here is receive distributor that is analogous to the sending side, so that implementations can do `receive_from(peer: Identity).await`).

Happy to put this on the TODO pile, of course.

---
OK, maybe I'm looking at this the wrong way around.  I was reacting to the asymmetry here, but I think perhaps I got it the wrong way around.  This is the thing that the networking piece will implement.  That leaves the ordering code as something that the networking code has to manage.  Ideally that is part of the generic functions eventually.

Again, more for the TODO pile.

Reply:
yea this is exactly we've been discussing with @thurstonsand. The reason it is implemented like presented here is that messaging layer has only one buffer for all messages (this seems to be desirable for asymmetric communication pattern we're observing). I think it is easy to change this to provide `peer: Identity` to the networking, but I can't think of a better strategy rather than random choice here, but I could be missing something. 

The pattern I am seeing on the protocol side is that we tend to send messages to one helper but expect to receive them from another, so activating a channel (`get_connection`) does not really mean that we will be receiving messages using it